cpp/Documentation.md

```diff
@@ -0,0 +1,474 @@
+# C++ Project Documentation
+
+This document provides instructions and hints to document C++ projects.
+
+## How to document code?
+
+### C++ API
+
+Public APIs may be documented with Doxygen, with comments specified in
+headers so that they remain visible to the developers consuming your APIs.
+
+### C++ Internal Code
+
+Documentation of internal code is highly recommended to help the maintainers
+of the project. Using Doxygen is not mandatory though. But if used, then the
+documentations of public and internal code should be well separated, such
+that it remains easy to the reader to distinguish public symbols from internal
+ones.
+
+Documentation of internal symbols can be specified in cpp files where symbols
+are defined, presumably more accessible to the maintainers.
+See [the following discussion](https://stackoverflow.com/questions/355619/where-to-put-the-doxygen-comment-blocks-for-an-internal-library-in-h-or-in-cpp)
```

Comment:
Actually this discussion contains another very useful point when is not discussed here: Splitting documentation of one symbol into the hpp and cpp file. After reading a few bits there this is what we could try:
- Document all API / public symbols in the hpp file with only the brief and short argument descriptions
- Add, when needed, additional documentation about the used algorithm, implementation, how to use or other technical details into the cpp file above the symbol's implementation
- Mark technical/implementation details such that it is clear in the generated html that these additional comments are only meant for the person who really wants to know more and they should appear last in the generated documentation for that symbol.

does that make sense?

Reply:
It makes perfect sense. See 53a76c9